File: src/core/Collector.ts
import { CollectorEvent, ResponsePayload, PricingConfig } from '../types';
import { store } from './Store';
import { eventBus } from './EventBus';
import { classifyError } from './classify';

const DEFAULT_PRICING: PricingConfig = {
    charsPerToken: 4,
    inputPerMillion: 3.0,
    outputPerMillion: 15.0,
};

const SECRET_KEYS = ['token', 'key', 'secret', 'password', 'auth', 'api_key',
    'apikey', 'credential', 'bearer', 'authorization'];

function sanitize(value: unknown, depth = 0): unknown {
    if (depth > 5) return value;
    if (typeof value !== 'object' || value === null) return value;
    if (Array.isArray(value)) return value.map(v => sanitize(v, depth + 1));

    const clean: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
        const isSecret = SECRET_KEYS.some(s => k.toLowerCase().includes(s));
        clean[k] = isSecret ? '[REDACTED]' : sanitize(v, depth + 1);
    }
    return clean;
}

function truncateResponse(raw: unknown): ResponsePayload {
    const serialized = JSON.stringify(raw);
    const sizeBytes = Buffer.byteLength(serialized, 'utf8');
    const LIMIT = 10_000;

    if (sizeBytes <= LIMIT) {
        return { data: raw, truncated: false, sizeBytes };
    }

    if (Array.isArray(raw)) {
        const kept: unknown[] = [];
        let size = 2;
        for (const item of raw) {
            const itemSize = Buffer.byteLength(JSON.stringify(item), 'utf8');
            if (size + itemSize > LIMIT) break;
            kept.push(item);
            size += itemSize + 1;
        }
        return { data: kept, truncated: true, sizeBytes };
    }

    if (typeof raw === 'object' && raw !== null) {
        const kept: Record<string, unknown> = {};
        for (const [k, v] of Object.entries(raw as Record<string, unknown>)) {
            kept[k] = typeof v === 'string' ? v.slice(0, 500) : v;
        }
        return { data: kept, truncated: true, sizeBytes };
    }

    return { data: String(raw).slice(0, LIMIT), truncated: true, sizeBytes };
}

class Collector {
    private pricing: PricingConfig = DEFAULT_PRICING;

    // Called once at dashboard startup (cli.ts `start`). Because every ingestion
    // path — multiplexer, per-server proxy, Python SDK — converges on this single
    // handle() in the dashboard process, configuring pricing here enriches all of
    // them. Pricing is intentionally a single source of truth in the dashboard.
    configure(pricing?: PricingConfig) {
        if (pricing) this.pricing = { ...this.pricing, ...pricing };
    }

    handle(event: CollectorEvent) {
        const sanitizedArgs = sanitize(event.arguments);
        const response = event.response
            ? (event.response.truncated !== undefined ? event.response : truncateResponse(event.response))
            : (event.response === null ? null : truncateResponse(event.response));

        const { inputTokens, outputTokens, costUsd } = this.estimateUsage(
            sanitizedArgs,
            response as ResponsePayload | null,
            event.serverName,
        );

        const processed: CollectorEvent = {
            ...event,
            arguments: sanitizedArgs,
            response: response as ResponsePayload | null,
            inputTokens,
            outputTokens,
            costUsd,
            errorClass: classifyError(event),
        };

        store.insertToolCall(processed);
        eventBus.emit('tool_call', processed);
    }

    // Estimate tokens from byte size. We deliberately use the ORIGINAL pre-truncation
    // size for output (response.sizeBytes) — truncation is a display/storage concern,
    // not an accounting one; the model still consumed the full payload.
    private estimateUsage(args: unknown, response: ResponsePayload | null, serverName: string) {
        const cpt = this.pricing.charsPerToken || 4;
        const argBytes = Buffer.byteLength(JSON.stringify(args ?? null), 'utf8');
        const outBytes = response?.sizeBytes ?? 0;

        const inputTokens = Math.ceil(argBytes / cpt);
        const outputTokens = Math.ceil(outBytes / cpt);

        const rate = this.pricing.perServer?.[serverName] ?? this.pricing;
        const costUsd =
            (inputTokens / 1_000_000) * rate.inputPerMillion +
            (outputTokens / 1_000_000) * rate.outputPerMillion;

        return { inputTokens, outputTokens, costUsd };
    }
}

export const collector = new Collector();

File: sdk/python/agent_monitor/collector.py
import threading
import uuid
import os
import json
from datetime import datetime, timezone
from typing import Any, Optional

MONITOR_URL = os.getenv("AGENT_MONITOR_URL", "http://localhost:4242")

_session_id = os.getenv("AGENT_MONITOR_SESSION_ID", str(uuid.uuid4()))

SECRET_KEYS = {'token', 'key', 'secret', 'password', 'auth', 'api_key',
               'apikey', 'credential', 'bearer', 'authorization'}


def _sanitize(value: Any, depth: int = 0) -> Any:
    if depth > 5:
        return value
    if isinstance(value, dict):
        return {
            k: '[REDACTED]' if any(s in k.lower() for s in SECRET_KEYS)
               else _sanitize(v, depth + 1)
            for k, v in value.items()
        }
    if isinstance(value, list):
        return [_sanitize(v, depth + 1) for v in value]
    return value


def _truncate_response(raw: Any) -> dict:
    serialized = json.dumps(raw)
    size_bytes = len(serialized.encode('utf-8'))
    limit = 10_000

    if size_bytes <= limit:
        return {"data": raw, "truncated": False, "sizeBytes": size_bytes}

    if isinstance(raw, list):
        kept, size = [], 2
        for item in raw:
            item_size = len(json.dumps(item).encode('utf-8'))
            if size + item_size > limit:
                break
            kept.append(item)
            size += item_size + 1
        return {"data": kept, "truncated": True, "sizeBytes": size_bytes}

    if isinstance(raw, dict):
        kept = {k: str(v)[:500] if isinstance(v, str) else v for k, v in raw.items()}
        return {"data": kept, "truncated": True, "sizeBytes": size_bytes}

    return {"data": str(raw)[:limit], "truncated": True, "sizeBytes": size_bytes}


def record(
    tool_name: str,
    server_name: str,
    arguments: Any,
    response: Any,
    status: str,
    latency_ms: float,
    timestamp: str,
    error: Optional[str] = None
):
    payload = {
        "sessionId": _session_id,
        "agentType": "python-sdk",
        "serverName": server_name,
        "toolName": tool_name,
        "method": tool_name,
        "arguments": _sanitize(arguments),
        "response": _truncate_response(response) if response is not None else None,
        "status": status,
        "latencyMs": round(latency_ms),
        "timestamp": timestamp,
        "errorMsg": error,
    }
    threading.Thread(target=_post, args=(payload,), daemon=True).start()


def _post(payload: dict):
    try:
        import urllib.request
        data = json.dumps(payload).encode('utf-8')
        req = urllib.request.Request(
            f"{MONITOR_URL}/api/ingest",
            data=data,
            headers={"Content-Type": "application/json"},
            method="POST"
        )
        urllib.request.urlopen(req, timeout=2)
    except Exception:
        pass
